File: codes/03-neoclassical-growth-model/julia/ncg_grid_search.jl
#=
SOLVING DETERMINISTIC NEOCLASSICAL GROWTH MODEL IN DISCRETE TIME USING GRID SEARCH

AUTHOR: Seho Jeong, Sogang University
DATE: 2025-09-14
REFERENCES:
- Edmond, Chris. 2019. "Lectures on Macroeconomics (PhD Core)." University of Melbourne.
- Hong, Jay H. 2025. "Lectures on Topics in Macroeconomics." Seoul National University.
=#

using LinearAlgebra
using Plots, LaTeXStrings
using BenchmarkTools

function NeoclassicalGrowth()
    # Set model parameters.
    β = 0.96 # time discount factor
    α = 0.4  # capital share in production function
    A = 1.0  # productivity
    δ = 1.0  # capital depreciation rate

    knum = 1001
    kmin = 1e-7
    kmax = 4.0

    kgrid = collect(range(kmin, kmax, knum))

    return (; β, α, A, δ, knum, kgrid)
end

# Solve Bellman equation by value function iteration.
function value_function_iteration(params; max_iter=1000, tol=1e-7, howard=false)

    # Unpack parameters.
    β, α, A, δ, knum, kgrid = params

    # Define utility and technology.
    u(c) = log(c)
    f(k) = A * k^α

    # Compute consumption matrix.
    c = f.(kgrid) .- (1 - δ) * kgrid .- kgrid'

    plt = heatmap(c, xlabel=L"k'", ylabel=L"k") # plot the matrix to check if it is computed appropriately.

    # Penalize violations of feasibility constraints.
    violations = c .<= 0
    c = c .* (c .>= 0) .+ 1e-10
    utility = u.(c) .- 1e16 * violations

    # Guess V.
    V = zeros(knum) # initial guess for value function
    g = zeros(knum)

    # Iterate on bellman operator.
    for iter in 1:max_iter
        
        # RHS of Bellman equation
        RHS = utility .+ β * V

        # Maximize over the RHS to get TV and the policy g.
        TV = maximum(RHS, dims=2)
        amax = argmax(RHS, dims=2)
        g = kgrid[amax]            # policy that attains the maximum

        # Check if converged.
        error = maximum(abs.(TV .- V))

        if mod(iter, 10) == 0
            println("Iteration $iter with error $error")
        end

        if error < tol
            println("Converged : iteration = $iter; error = $error")
            break
        end

        V = copy(TV)
    end

    return V, g
end

params = NeoclassicalGrowth()
@time V, g = value_function_iteration(params)

# Plot the results
result1 = plot(kgrid, V, xlabel=L"k", ylabel=L"V(k)")

File: codes/03-neoclassical-growth-model/matlab/ncg_grid_search.m
%{
SOLVING DETERMINISTIC NEOCLASSICAL GROWTH IN DISCRETE TIME WITH GRID SEARCH

AUTHOR: Seho Jeong, Sogang University
DATE: 2025-09-14
REFERENCES:
- Edmond, Chris. 2019. "Lectures on Macroeconomics (PhD Core)." University of Melbourne.
- Hong, Jay H. 2025. "Lectures on Topics in Macroeconomics." Seoul National University.
%}

clear all; clc

% Model parameters
alpha = 1/3;        % capital's share in production function
beta  = 0.95;       % time discount factor
delta = 0.05;       % depreciation rate
sigma = 1;          % CRRA (=1/IES)
rho   = (1/beta)-1; % implied rate of time preference

kstar = (alpha / (rho + delta))^(1/(1-alpha)); % steady state
kbar  = (1/delta)^(1/(1-alpha));

% Numerical parameters
max_iter = 500;   % maximum number of iterations
tol      = 1e-7;  % treat numbers smaller than this as zero
penalty  = 10^16; % for penalizing constraint violations

% Set up the grid of capital stocks.
knum = 1001; % number of nodes for k grid
kmin = tol;  % effectively zero
kmax = kbar; % effective upper bound on k grid

kgrid = linspace(kmin, kmax, knum); % linearly spaced

% Return function
c = zeros(knum, knum);
for j=1:knum,
    c(:, j) = (kgrid.^alpha) + (1 - delta)*kgrid - kgrid(j);    
end

% Penalize violations of feasibility constraints.
violations = (c <= 0);
c = c .* (c >= 0) + eps;

if sigma == 1,
    u = log(c) - penalty * violations;
else
    u = (1 / (1 - sigma)) * (c.^(1 - sigma) - 1) - penalty * violations;
end

% Solve Bellman equation by value function iteration.

% Initial guess
V = zeros(knum, 1);

hold on
% Iterate on Bellman operator
for iter=1:max_iter
    % RHS of Bellman equation
    RHS = u + beta * kron(ones(knum, 1), V');

    % Maximize over this to get Tv.
    [TV, argmax] = max(RHS, [], 2);

    % Policy that attains the maximum
    g = kgrid(argmax);

    % Check if converged.
    error = norm(TV - V, inf);
    if error < tol
        fprintf('%4i %6.2e \n', [iter, error])
        break
    end

    V = TV;
    if rem(iter, 10) == 0
        fprintf('%4i %6.2e \n', [iter, error]);
    end
end

plot(kgrid, V)
hold off


% plot(kgrid, g)